File: c/quick3.c
#include <stdlib.h>

void swap(int *a, int *b);
int min(int a, int b);

// quick3 implements quick sort (3 way) algorithm. The function takes an array and its length.
void quick3(int a[], int n) {
	if(n <= 1) {
		return;
	}
	swap(&a[n-1], &a[(int)(rand())%n]);
	int i = 0, k = 0, p = n;
	while(i < p) {
		if(a[i] < a[n-1])
			swap(&a[i++], &a[k++]);
		else if(a[i] == a[n-1])
			swap(&a[i], &a[--p]);
		else
			i++;
	}
	int m = min(p-k, n-p+1);
	for(int i = 0; i < m-1; i++)
		swap(&a[k+i], &a[n-m+1+i]);
	quick3(a, k);
	quick3(a+k+1, n-k-1);
}

File: php/quick3.php
<?php
require_once('util.php');

// quick3 implements quick sort (3 way) algorithm. The function takes an array.
function quick3(&$a) {
	_quick3($a, 0, count($a));
}

function _quick3(&$a, $l, $n) {
	if($n <= 1)
		return;
	swap($a, $l+$n-1, $l+rand(0, $n-1));
	$i = $l;
	$k = $l;
	$p = $l+$n;
	while($i < $p) {
		if($a[$i] < $a[$l+$n-1])
			swap($a, $i++, $k++);
		else if($a[$i] == $a[$l+$n-1])
			swap($a, $i, --$p);
		else
			$i++;
	}
	$m = min($p-$k, $l+$n-$p+1);
	for($i = 0; $i < $m-1; $i++)
		swap($a, $k+$i, $l+$n-$m+1+$i);
	_quick3($a, $l, $k-$l);
	_quick3($a, $k+1, $l+$n-$k-1);
}
